Lombiq.Tests.UI/Extensions/IWebApplicationInstanceExtensions.cs

```diff
@@ -22,27 +35,71 @@ public static class IWebApplicationInstanceExtensions
         bool activateShell = true)
     {
         var shellHost = instance.GetRequiredService<IShellHost>();
-        // Injecting a fake HttpContext is required to avoid NullReferenceException in
-        // OrchardCore.Recipes.Services.RecipeEnvironmentFeatureProvider.PopulateEnvironmentAsync.
+
         var httpContextAccessor = instance.GetRequiredService<IHttpContextAccessor>();
-        var features = new FeatureCollection();
-        features.Set(new RecipeEnvironmentFeature());
+        var originalHttpContext = httpContextAccessor.HttpContext;
 
-        var httpContext = new DefaultHttpContext(features)
+        try
         {
-            RequestServices = instance.Services,
-        };
+            // Injecting a fake HttpContext is required for many things, but it needs to happen before UsingAsync()
+            // below to avoid NullReferenceExceptions in
+            // OrchardCore.Recipes.Services.RecipeEnvironmentFeatureProvider.PopulateEnvironmentAsync. Migrations
+            // (possibly with recipe migrations) run right at the shell start.
 
-        httpContextAccessor.HttpContext = httpContext;
+            var shellScope = await shellHost.GetScopeAsync(tenant);
 
-        try
-        {
-            await (await shellHost.GetScopeAsync(tenant))
-                .UsingAsync(execute, activateShell);
+            // Creating a fake HttpContext like in ModularBackgroundService.
+            httpContextAccessor.HttpContext = shellScope.ShellContext.CreateHttpContext();
+            var httpContext = httpContextAccessor.HttpContext;
+
+            httpContext.Request.PathBase = "/" + shellHost.GetSettings(tenant).RequestUrlPrefix ?? string.Empty;
+            httpContext.Features.Set(new RecipeEnvironmentFeature());
+
+            await shellScope.UsingAsync(execute, activateShell);
         }
         finally
         {
-            httpContextAccessor.HttpContext = null;
+            httpContextAccessor.HttpContext = originalHttpContext;
+        }
+    }
+}
+
+// This is from the 1.4 version of Orchard Core. During an Orchard upgrade to 1.5, remove this, and use the now public
+// implementation in OrchardCore.Abstractions.
```

Comment:
Not really, since our OC upgrade guide already tells you to search for, among others "Orchard upgrade". The points under that issue are rather things that aren't this obvious.

Reply:
I'm not too happy about that, because that's just a todo comment by another name. Seems like a somewhat fragile format too, because if the "Orchard upgrade" text is caught by a line break (maybe automatically) then search won't be able to fine it later.
It's not a problem here, so I'm not going start a revolution over it... for now!

Reply:
That wouldn't be a problem if we always maintain a single vNext issue (so create a new one once the latest is resolved) in the OSOCE repo, and link it in the https://lombiq.atlassian.net/wiki/spaces/OSOE/pages/75137069/Developer+overview

Reply:
Yeah, but then you have to also keep it updated :). I mean, it's not less work or less fragile, and you'd still need to link the issue from the code for clarity. The line break issue should be possible to work around with a regex search.

Reply:
I don't get how it's no less fragile tbh, a shared list of all todos sounds pretty robust.

Reply:
You have to know that the issue exists and you have to update it. It moves information about the code from the code to an issue. Since there can be a lot of time between us noticing an issue that'll be fixed by an OC upgrade (or that needs special care during it) and when the upgrade happens, it's safer to keep it in the code, as it relies less on people knowing about things that are less noticable.

Reply:
Judging by previous comments of yours, Zoltán, on other TODO-ish comments, I feel like this one should also be moved into an issue, much like Dávid suggested in his first comment. I don't really understand the difference between this and other situations. If we know of this todo today, let's put it on an explicit list in the update issue and be done with it, I would say.

Reply:
Really, as I explained above. These are not just random TO DO comments but part of an organized process of documenting areas of the code that need special care during an Orchard upgrade (this can BTW be continuous too, like the NLog version in the UI Testing Toolbox, not just a one-off task). These are not "do this when you have time" kind of tasks but rather "if you change this code in a particular way in the future, then do that".